repos/system_upgrade/common/libraries/overlaygen.py

```diff
@@ -128,29 +284,98 @@ def cleanup_scratch(scratch_dir, mounts_dir):
     api.current_logger().debug('Recursively removed scratch directory %s.', scratch_dir)
 
 
-def _create_mount_disk_image(disk_images_directory, path):
+def _format_disk_image_ext4(diskimage_path):
     """
-    Creates the mount disk image, for cases when we hit XFS with ftype=0
+    Format the specified disk image with Ext4 filesystem.
+
+    The formatted file system is optimized for operations we want to do and
+    mainly for the space it needs to take for the initialisation. So use 32MiB
+    journal (that's enough for us as we do not plan to do too many operations
+    inside) for any size of the disk image. Also the lazy
+    initialisation is disabled. The formatting will be slower, but it helps
+    us to estimate better the needed amount of the space for other actions
+    done later.
     """
-    diskimage_path = os.path.join(disk_images_directory, _mount_name(path))
-    disk_size = _overlay_disk_size()
-
-    api.current_logger().debug('Attempting to create disk image with size %d MiB at %s', disk_size, diskimage_path)
-    utils.call_with_failure_hint(
-        cmd=['/bin/dd', 'if=/dev/zero', 'of={}'.format(diskimage_path), 'bs=1M', 'count={}'.format(disk_size)],
-        hint='Please ensure that there is enough diskspace in {} at least {} MiB are needed'.format(
-            diskimage_path, disk_size)
-    )
-
     api.current_logger().debug('Creating ext4 filesystem in disk image at %s', diskimage_path)
+    cmd = [
+        '/sbin/mkfs.ext4',
+        '-J', 'size=32',
+        '-E', 'lazy_itable_init=0,lazy_journal_init=0',
+        '-F', diskimage_path
+    ]
     try:
-        utils.call_with_oserror_handled(cmd=['/sbin/mkfs.ext4', '-F', diskimage_path])
+        utils.call_with_oserror_handled(cmd=cmd)
     except CalledProcessError as e:
+        # FIXME(pstodulk): taken from original, but %s seems to me invalid here
         api.current_logger().error('Failed to create ext4 filesystem %s', exc_info=True)
```

Comment:
Agree. That `%s` should not be there, or the `diskimage_path` should be passed to the logging call.